addon/queue.ts

```diff
@@ -0,0 +1,281 @@
+      listener.fileAdded?.(file);
+    }
+  }
+
+  /**
+   * Remove a file from the queue
+   * @param file the file to be removed
+   */
+  @action
+  remove(file: UploadFile) {
+    if (!this.#distinctFiles.has(file)) {
+      return;
+    }
+
+    file.queue = undefined;
+    this.#distinctFiles.delete(file);
+
+    for (const listener of this.#listeners) {
+      listener.fileRemoved?.(file);
+    }
+  }
+
+  /**
+    @private
+    @method _addFiles
+    @param {FileList} fileList The event triggered from the DOM that contains a list of files
+   */
+  _addFiles(fileList: FileList, source: FileSource) {
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore
+    const onFileAdd = this.onFileAdd;
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore
+    const disabled = this.disabled;
+    const files: UploadFile[] = [];
+
+    if (!disabled) {
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      // @ts-ignore
+      for (let i = 0, len = fileList.length || fileList.size; i < len; i++) {
+        const file = fileList.item ? fileList.item(i) : fileList[i];
+        if (file instanceof File) {
+          const uploadFile = new UploadFile(file, source);
+
+          files.push(uploadFile);
+          this.push(uploadFile);
+
+          if (onFileAdd) {
+            next(onFileAdd, uploadFile);
+          }
+        }
+      }
+    }
+
+    return files;
+  }
+
+  /**
+   * Flushes the `files` property if they have settled. This
+   * will only flush files when all files have arrived at a terminus
+   * of their state chart (`uploaded` and `aborted`).
+   *
+   * Files *may* be requeued by the user in the `failed` or `timed_out`
+   * states.
+   */
+  flush() {
+    if (this.files.length === 0) {
+      return;
+    }
+
+    const allFilesHaveSettled = this.files.every((file) => {
+      return [FileState.Uploaded, FileState.Aborted].includes(file.state);
+    });
+
+    if (allFilesHaveSettled) {
+      this.files.forEach((file) => (file.queue = undefined));
+      this.#distinctFiles.clear();
+    }
+  }
+
+  selectFile = modifier<
+    HTMLInputElement,
+    SelectFileModifierArgs['positional'],
+    SelectFileModifierArgs['named']
+  >((element, _positional, named) => {
+    const changeHandler = (event: Event) => {
+      const { files } = event.target as HTMLInputElement;
+      if (!files) {
+        return;
+      }
+
+      const selectedFiles: UploadFile[] = [];
+
+      for (const file of files) {
+        if (named.filter && !named.filter?.(file)) {
+          continue;
+        }
+
+        let uploadFile;
+        if (file instanceof File) {
+          uploadFile = new UploadFile(file, FileSource.Browse);
+        }
+        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+        // @ts-ignore
+        else if (file instanceof Blob) {
+          uploadFile = UploadFile.fromBlob(file, FileSource.Browse);
+        }
```

Comment:
@gilest I made this case-switch after your changes on `UploadFile` constructor. That was necessary to make `selectFiles()` from `tests-support` work.

Is this as you intended it? Wdyt?